dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java

```diff
@@ -92,114 +91,74 @@ public class EE9FeatureCompatibilityTest extends FATServletClient {
 
-        incompatibleValueAddFeatures.add("openid-2.0"); // stabilized
-        incompatibleValueAddFeatures.add("openapi-3.1"); // depends on mpOpenAPI
-        incompatibleValueAddFeatures.add("opentracing-1.0"); // opentracing depends on mpConfig
-        incompatibleValueAddFeatures.add("opentracing-1.1");
-        incompatibleValueAddFeatures.add("opentracing-1.2");
-        incompatibleValueAddFeatures.add("opentracing-1.3");
-        incompatibleValueAddFeatures.add("opentracing-2.0");
-        incompatibleValueAddFeatures.add("sipServlet-1.1"); // purposely not supporting EE 9
-        incompatibleValueAddFeatures.add("springBoot-1.5"); // springBoot 3.0 will support EE 9
-        incompatibleValueAddFeatures.add("springBoot-2.0");
+        // Test features may or may not be compatible, we don't want to assert either way
+        compatibleFeatures.removeAll(FeatureUtilities.allTestFeatures());
+
+        return compatibleFeatures;
     }
 
-    static Set<String> getAllCompatibleFeatures() {
-        Set<String> compatFeatures = new HashSet<>();
-        try {
-            File featureDir = new File(Bootstrap.getInstance().getValue("libertyInstallPath") + "/lib/features/");
-            // If there was a problem building projects before this test runs, "lib/features" won't exist
-            if (featureDir != null && featureDir.exists()) {
-                for (File feature : featureDir.listFiles()) {
-                    if (feature.getName().startsWith("io.openliberty.") ||
-                        feature.getName().startsWith("com.ibm.")) {
-                        String shortName = EE9FeatureCompatibilityTest.parseShortName(feature);
-                        if (shortName != null) {
-                            compatFeatures.add(shortName);
-                        }
-                    }
-                }
-            }
-        } catch (Exception e) {
-            throw new RuntimeException(e);
-        }
-        compatFeatures.removeAll(EE9FeatureCompatibilityTest.nonEE9JavaEEFeatures);
-        compatFeatures.removeAll(EE9FeatureCompatibilityTest.nonEE9MicroProfileFeatures);
-        compatFeatures.removeAll(EE9FeatureCompatibilityTest.incompatibleValueAddFeatures);
+    public static Set<String> getIncompatibleFeatures(Set<String> allFeatures, Set<String> compatibleFeatures) {
+        Set<String> incompatibleFeatures = new HashSet<>();
 
-        // when concurrent-3.0 moves to EE10 only you can remove this line since it won't be in the Set any longer.
-        compatFeatures.remove("concurrent-3.0");
+        // Logically, incompatible features are all those that aren't compatible...
+        incompatibleFeatures.addAll(allFeatures);
+        incompatibleFeatures.removeAll(compatibleFeatures);
 
-        return compatFeatures;
+        // ...but there are a few in-development features which are maybe compatible
+        incompatibleFeatures.removeAll(MicroProfileActions.MP60.getFeatures());
```

Comment:
Is this really needed?

Reply:
It is at the moment. `mpOpenAPI-3.1` and `mpTelemetry-1.0` are currently compatible with EE9, though they will not be by the time they're ready to release.